test/metabase/util_test.cljc

```diff
@@ -470,11 +472,20 @@
     (is (= [:c :e] (u/conflicting-keys {:a 1 :b 2 :c 3 :e nil}
                                        {:b 2 :c 4 :d 5 :e 6})))))
 
+(defn- join [& crumbs] (str/join ":" crumbs))
+
 (deftest map-all-test
-  (let [join (fn [& crumbs] (str/join ":" crumbs))]
-    (testing "map-all with 1 collection is just map"
-      (is (= ["0" "1" "2" "3" "4"] (u/map-all join (range 5)))))
-    (testing "map-all works with 3 collections"
-      (is (= ["0:0" "1:1" "2:2" "3:" "4:"] (u/map-all join (range 5) (range 3)))))
-    (testing "map-all works with higher arity"
-      (is (= ["0:0:0" "1:1:1" "2:2:2" "3::3" "::4"] (u/map-all join (range 4) (range 3) (range 5)))))))
+  (testing "map-all with 1 collection is just map"
+    (is (= ["0" "1" "2" "3" "4"] (u/map-all join (range 5)))))
+  (testing "map-all works with 3 collections"
+    (is (= ["0:0" "1:1" "2:2" "3:" "4:"] (u/map-all join (range 5) (range 3)))))
+  (testing "map-all works with higher arity"
+    (is (= ["0:0:0" "1:1:1" "2:2:2" "3::3" "::4"] (u/map-all join (range 4) (range 3) (range 5))))))
+
+#?(:clj
+   (defspec map-all-returns-max-coll-input-size-test 1000
```

Comment:
Some other invariant that pop into my head:

- depends on f / consistent with map: `(map f ...)` == `(take n (map-all f ...))` where `n` = min count
- depends on every value in every collection: `(map-all (fnil + 0 0 0) ...)` == `(sum (map sum colls))`

Feels like together with your invariant this would give pretty good coverage.

.. there's also the option to just compare to a naive implementation, but slippery if it's complex enough to have its own bugs.

```clojure
(let [n   (apply max (map count colls))
      pad #(take n (concat % (repeat nil)))]
  (apply map f (map pad colls)))
```

Reply:
That fnil example is a good idea, I'll write that one in too.

Reply:
> naive implementation, but slippery if it's complex enough to have its own bugs.

Yeah, I have found that it can be tricky to specify complicated constraints without re-writing the code.